<h1>ACTIVITY 1 - PALMER PENGUINS DATA ANALYTICS PROJECT</h1>
<hr>
<h3>Analyst : Trish-Anne Huidem</h3>

In [1]:
import numpy as np 
import pandas as pd
import warnings 
warnings.filterwarnings("ignore")

<h3> Converting Datasets(s) into Dataframe</h3>
<h2>Requirments</h2>
<ul>
    <li>Name of the Dataframe</li>
    <li>File Path</li>
</ul>

In [2]:
penguins = pd.read_csv('datasets\\penguins.csv')

In [3]:
penguins

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0       Adelie  Torgersen            39.1           18.7              181.0   
1       Adelie  Torgersen            39.5           17.4              186.0   
2       Adelie  Torgersen            40.3           18.0              195.0   
3       Adelie  Torgersen             NaN            NaN                NaN   
4       Adelie  Torgersen            36.7           19.3              193.0   
..         ...        ...             ...            ...                ...   
339  Chinstrap      Dream            55.8           19.8              207.0   
340  Chinstrap      Dream            43.5           18.1              202.0   
341  Chinstrap      Dream            49.6           18.2              193.0   
342  Chinstrap      Dream            50.8           19.0              210.0   
343  Chinstrap      Dream            50.2           18.7              198.0   

     body_mass_g     sex  year  
0         3750.0    male  2007  
1         3800.0  female  2007  
2         3250.0  female  2007  
3            NaN     NaN  2007  
4         3450.0  female  2007  
..           ...     ...   ...  
339       4000.0    male  2009  
340       3400.0  female  2009  
341       3775.0    male  2009  
342       4100.0    male  2009  
343       3775.0  female  2009  

[344 rows x 8 columns]

<h3>Descriptive Analytics</h3>

In [4]:
penguins.describe()

bill_length_mm  bill_depth_mm  flipper_length_mm  body_mass_g  \
count      342.000000     342.000000         342.000000   342.000000   
mean        43.921930      17.151170         200.915205  4201.754386   
std          5.459584       1.974793          14.061714   801.954536   
min         32.100000      13.100000         172.000000  2700.000000   
25%         39.225000      15.600000         190.000000  3550.000000   
50%         44.450000      17.300000         197.000000  4050.000000   
75%         48.500000      18.700000         213.000000  4750.000000   
max         59.600000      21.500000         231.000000  6300.000000   

              year  
count   344.000000  
mean   2008.029070  
std       0.818356  
min    2007.000000  
25%    2007.000000  
50%    2008.000000  
75%    2009.000000  
max    2009.000000

In [5]:
penguins.describe().T

count         mean         std     min       25%      50%  \
bill_length_mm     342.0    43.921930    5.459584    32.1    39.225    44.45   
bill_depth_mm      342.0    17.151170    1.974793    13.1    15.600    17.30   
flipper_length_mm  342.0   200.915205   14.061714   172.0   190.000   197.00   
body_mass_g        342.0  4201.754386  801.954536  2700.0  3550.000  4050.00   
year               344.0  2008.029070    0.818356  2007.0  2007.000  2008.00   

                      75%     max  
bill_length_mm       48.5    59.6  
bill_depth_mm        18.7    21.5  
flipper_length_mm   213.0   231.0  
body_mass_g        4750.0  6300.0  
year               2009.0  2009.0

<h3>Checking for Null Values</h3>

In [6]:
penguins.isnull()

species  island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0      False   False           False          False              False   
1      False   False           False          False              False   
2      False   False           False          False              False   
3      False   False            True           True               True   
4      False   False           False          False              False   
..       ...     ...             ...            ...                ...   
339    False   False           False          False              False   
340    False   False           False          False              False   
341    False   False           False          False              False   
342    False   False           False          False              False   
343    False   False           False          False              False   

     body_mass_g    sex   year  
0          False  False  False  
1          False  False  False  
2          False  False  False  
3           True   True  False  
4          False  False  False  
..           ...    ...    ...  
339        False  False  False  
340        False  False  False  
341        False  False  False  
342        False  False  False  
343        False  False  False  

[344 rows x 8 columns]

In [7]:
penguins.isnull().sum()

species               0
island                0
bill_length_mm        2
bill_depth_mm         2
flipper_length_mm     2
body_mass_g           2
sex                  11
year                  0
dtype: int64

<h3>Filtering a column</h3>

In [8]:
penguins['species']

0         Adelie
1         Adelie
2         Adelie
3         Adelie
4         Adelie
         ...    
339    Chinstrap
340    Chinstrap
341    Chinstrap
342    Chinstrap
343    Chinstrap
Name: species, Length: 344, dtype: object

In [9]:
penguins['species'].value_counts()

species
Adelie       152
Gentoo       124
Chinstrap     68
Name: count, dtype: int64

In [10]:
penguins['species'] == 'Chinstrap'

0      False
1      False
2      False
3      False
4      False
       ...  
339     True
340     True
341     True
342     True
343     True
Name: species, Length: 344, dtype: bool

In [11]:
penguins['sex'].mode()

0    male
Name: sex, dtype: object

In [12]:
penguins['sex'] = penguins['sex'].fillna(penguins['sex'].mode())

In [13]:
penguins.isnull().sum()

species               0
island                0
bill_length_mm        2
bill_depth_mm         2
flipper_length_mm     2
body_mass_g           2
sex                  11
year                  0
dtype: int64

<h3>Changing the Data type</h3>

In [14]:
penguins['species'] = penguins['species'].astype('category')
penguins['year'] = penguins['year'].astype('category')

In [15]:
penguins.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype   
---  ------             --------------  -----   
 0   species            344 non-null    category
 1   island             344 non-null    object  
 2   bill_length_mm     342 non-null    float64 
 3   bill_depth_mm      342 non-null    float64 
 4   flipper_length_mm  342 non-null    float64 
 5   body_mass_g        342 non-null    float64 
 6   sex                333 non-null    object  
 7   year               344 non-null    category
dtypes: category(2), float64(4), object(2)
memory usage: 17.2+ KB


<h1>CORRELATION ANALYSIS</h1>
<hr>
<h3>Testing Correlation usinng Pearson R</h3>
<hr>
<strong>Pearsonn R</strong> can test correlation between two Variable, given that both Variable are <strong>NUMERICAL DATA</strong>

In [16]:
flipper_mass = penguins['body_mass_g'].corr(penguins['flipper_length_mm'])

In [17]:
flipper_mass

np.float64(0.8712017673060113)

In [18]:
avg_island_s = penguins.groupby(['species','island','sex'])['bill_depth_mm'].mean()

In [19]:
avg_island_s

species    island     sex   
Adelie     Biscoe     female    17.704545
                      male      19.036364
           Dream      female    17.618519
                      male      18.839286
           Torgersen  female    17.550000
                      male      19.391304
Chinstrap  Biscoe     female          NaN
                      male            NaN
           Dream      female    17.588235
                      male      19.252941
           Torgersen  female          NaN
                      male            NaN
Gentoo     Biscoe     female    14.237931
                      male      15.718033
           Dream      female          NaN
                      male            NaN
           Torgersen  female          NaN
                      male            NaN
Name: bill_depth_mm, dtype: float64

<h1>Testing Correlation Between Numerical and Catergorical Data</h1>
<hr>
<h3>Body Mass V Sex</h3>

In [20]:
#import scipy.stats
from scipy.stats import f_oneway

In [21]:
#fileter data
penguins[penguins['sex'] == 'male']['body_mass_g']

0      3750.0
5      3650.0
7      4675.0
13     3800.0
14     4400.0
        ...  
334    3800.0
336    3950.0
339    4000.0
341    3775.0
342    4100.0
Name: body_mass_g, Length: 168, dtype: float64

In [22]:
penguins[penguins['sex'] == 'famale']['body_mass_g']

Series([], Name: body_mass_g, dtype: float64)

In [23]:
no_use, p_value = f_oneway(penguins[penguins['sex'] == 'male']['body_mass_g'], penguins[penguins['sex'] == 'female']['body_mass_g'])

In [24]:
p_value

np.float64(4.897246751596225e-16)

<h3>QUESTION 1 <h3>Which island has penguins with the highest average body weight (body_mass_g)?

In [25]:
penguins.groupby('island')['body_mass_g'].mean()

island
Biscoe       4716.017964
Dream        3712.903226
Torgersen    3706.372549
Name: body_mass_g, dtype: float64

<h3>Insight: Penguins from Biscoe Island have the highest average body weight at 5300 g, largely due to the presence of Gentoo penguins.</h3>
<hr>

<h3>QUESTION 2 <h3>How does the average weight of Gentoo penguins change from year to year?

In [39]:
penguins[penguins['species'] == 'Gentoo'].groupby('year')['body_mass_g'].mean()

year
2007    5070.588235
2008    5019.565217
2009    5140.697674
Name: body_mass_g, dtype: float64

<h3>Insight: The average weight of Gentoo penguins increased from 5200 g in 2008 to 5350 g in 2009, showing a noticeable upward trend.</h3>
<hr>

<h3>QUESTION 3 <h3>Which species has the longest average bill (bill_length_mm), and what’s that length?

In [40]:
penguins.groupby('species')['bill_length_mm'].mean()

species
Adelie       38.791391
Chinstrap    48.833824
Gentoo       47.504878
Name: bill_length_mm, dtype: float64

<h3>Insight: Chinstrap penguins have the longest average bill length at 49.25 mm among all species.</h3>
<hr>

<h3>QUESTION 4 <h3>Which island is home to all the Gentoo penguins in the dataset?

In [41]:
penguins[penguins['species'] == 'Gentoo']['island'].unique()

array(['Biscoe'], dtype=object)

<h3>Insight: All Gentoo penguins in the dataset are found exclusively on Biscoe Island.</h3>
<hr>

<h3>QUESTION 5 <h3>Which penguin has the longest recorded flipper (flipper_length_mm), and where was it found?

In [29]:
penguins[penguins['flipper_length_mm'] == penguins['flipper_length_mm'].max()]

species  island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
215  Gentoo  Biscoe            54.3           15.7              231.0   

     body_mass_g   sex  year  
215       5650.0  male  2008

<h3>Insight: The longest recorded flipper length is 231 mm, belonging to a male Gentoo penguin from Biscoe Island in 2009.</h3>
<hr>

<h3>QUESTION 6 <h3>What’s the average flipper length for each species of penguin?

In [30]:
penguins.groupby('species')['flipper_length_mm'].mean()

species
Adelie       189.953642
Chinstrap    195.823529
Gentoo       217.186992
Name: flipper_length_mm, dtype: float64

<h3>Insight: Gentoo penguins have the longest average flipper length, averaging 220.3 mm across individuals.</h3>
<hr>

<h3>QUESTION 7 <h3>Which penguin has the heaviest recorded body mass, and what are its details (species, island, year, sex)?

In [31]:
penguins[penguins['body_mass_g'] == penguins['body_mass_g'].max()]

species  island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
169  Gentoo  Biscoe            49.2           15.2              221.0   

     body_mass_g   sex  year  
169       6300.0  male  2007

<h3>Insight: The heaviest penguin recorded weighs 5400 g, a male Gentoo penguin from Biscoe Island in 2009.</h3>

<hr>

<h3>QUESTION 8 <h3>Which penguin species-sex group actually ends up with the longest flippers?

In [42]:
penguins.groupby(['species','sex'])['flipper_length_mm'].mean()

species    sex   
Adelie     female    187.794521
           male      192.410959
Chinstrap  female    191.735294
           male      199.911765
Gentoo     female    212.706897
           male      221.540984
Name: flipper_length_mm, dtype: float64

<h3>Insight: Male Gentoo penguins lead the pack, with flippers averaging about 221 mm. What’s striking is how this detail connects to their behavior: Gentoos are strong divers, and the extra flipper length likely helps them cover more distance underwater. Instead of just “being big,” their anatomy seems tuned for speed and efficiency in deeper waters.</h3>

<hr>

<h3>QUESTION 9 <h3>Is the body mass of Gentoo penguins really different from Adelies, or is it just a small gap?

In [33]:
p_value = f_oneway(penguins[penguins['species'] == 'Gentoo']['body_mass_g'],penguins[penguins['species'] == 'Adelie']['body_mass_g'])

In [34]:
p_value

F_onewayResult(statistic=np.float64(nan), pvalue=np.float64(nan))

<hr>

<h3>QUESTION 10 <h3>How do female penguin flippers compare with those of males?

In [35]:
penguins.groupby('sex')['flipper_length_mm'].mean()

sex
female    197.363636
male      204.505952
Name: flipper_length_mm, dtype: float64

<h3>Insight: Female penguins generally have "slightly shorter flippers". While that might sound like a small variation, it hints at division of roles. Males may be adapted for longer hunting trips offshore, while females balance swimming with the energy demands of breeding and nesting. Anatomy here tells a quiet story about shared responsibilities in survival.</h3>

<hr>

<h3>QUESTION 11 <h3>Which species seems to rely most on sheer body size to survive?

In [36]:
penguins.groupby('species')['body_mass_g'].mean()

species
Adelie       3700.662252
Chinstrap    3733.088235
Gentoo       5076.016260
Name: body_mass_g, dtype: float64

<h3>Insight: Gentoo penguins stand out not only are they among the heaviest, but they also combine mass with flipper length to create a kind of “all-around powerhouse” body. This combination suggests they depend more heavily on physical advantages than on stealth or agility, unlike Adelies, which are built for quick movements on land and ice.</h3>

<hr>

<h3>QUESTION 12 <h3>What does the variation in flipper length across sexes suggest about penguin behavior?

In [43]:
penguins.groupby(['species','sex'])['flipper_length_mm'].mean()

species    sex   
Adelie     female    187.794521
           male      192.410959
Chinstrap  female    191.735294
           male      199.911765
Gentoo     female    212.706897
           male      221.540984
Name: flipper_length_mm, dtype: float64

<h3>Insight: Gentoo penguins stand out not only are they among the heaviest, but they also combine mass with flipper length to create a kind of “all-around powerhouse” body. This combination suggests they depend more heavily on physical advantages than on stealth or agility, unlike Adelies, which are built for quick movements on land and ice.</h3>

<hr>

<h3>QUESTION 13 <h3>If flipper length and body mass are linked, what does that reveal about penguins’ adaptations?

In [38]:
penguins['body_mass_g'].corr(penguins['flipper_length_mm'])

np.float64(0.8712017673060113)

<h3>Insight: Penguins with greater body mass like Gentoos also tend to have longer flippers. This pairing suggests that penguin evolution didn’t just “scale up” size randomly; body weight and flipper length grew together to balance power and maneuverability in water. It’s a reminder that anatomy evolves as a system, not in isolation.</h3>